kafka-client/src/main/java/dev/responsive/kafka/store/ResponsivePartitionedStore.java

```diff
@@ -130,22 +142,100 @@ public void init(final StateStoreContext context, final StateStore root) {
           new BytesKeySpec(),
           config
       );
-      buffer.init();
 
-      open = true;
+      if (isActive()) {
+        registration = new ResponsiveStoreRegistration(
+            name.kafkaName(),
+            topicPartition,
+            buffer::flush,
+            this::storedOffset
+        );
+      } else {
+        registration = new ResponsiveStoreRegistration(
+            name.kafkaName(),
+            topicPartition,
+            buffer::flush,
+            this::storedOffset,
+            this::maybeTransitionToActive
+        );
+      }
 
-      context.register(root, buffer);
-      final long offset = buffer.offset();
-      registration = new ResponsiveStoreRegistration(
-          name.kafkaName(),
-          topicPartition,
-          offset == -1 ? 0 : offset,
-          buffer::flush
-      );
-      storeRegistry = InternalConfigs.loadStoreRegistry(context.appConfigs());
       storeRegistry.registerStore(registration);
+
+      open = true;
+      context.register(
+          root,
+          new ActiveFilteringRestoreCallback((batch) -> buffer.restoreBatch(batch), this::isActive)
+      );
     } catch (InterruptedException | TimeoutException e) {
-      throw new ProcessorStateException("Failed to initialize store.", e);
+      throw new ProcessorStateException("Failed to initialize store " + name(), e);
+    }
+  }
+
+  private boolean isActive() {
+    return context.taskType() == TaskType.ACTIVE;
+  }
+
+  /**
+   * The current stored offset in the remote table. Guaranteed to be less than or equal to the
+   * committed offset of the corresponding changelog topic partition in Kafka
+   * <p>
+   * NOTE: this always makes a remote call, should be cached eg when used to obtain starting offset
```

Comment:
I think it's better not to cache this as it can be updated by other writers - it's a pretty lightweight call so unless it's called in a tight loop we should be good

Reply:
Well right now it's only ever used to initialize the offset at the start of restoration -- in the current code it's essentially cached as the "committed offset" in ResponsiveStoreRegistration. But yeah, if you actually need the updated current stored offset, you should definitely not use a cached value 🙂